src/host/_stream.cpp

```diff
@@ -126,565 +165,286 @@ void AdjustCursorPosition(SCREEN_INFORMATION& screenInfo, _In_ til::point coordC
+
+                        // Now that we finally know how many columns precede the cursor we can
+                        // subtract the previously determined amount of ' ' from the '\t'.
+                        glyphCount -= gsl::narrow_cast<til::CoordType>(backupEnd - backupBeg);
+
+                        // Can the above code leave glyphCount <= 0? Let's just not find out!
+                        glyphCount = std::max(1, glyphCount);
                     }
-                    else
+                    // Control chars in interactive mode where previously written out
+                    // as ^X for instance, so now we also need to delete 2 glyphs.
+                    else if (IS_CONTROL_CHAR(lastChar))
                     {
-                        FAIL_FAST_IF(!(Tmp2 >= buffer.get()));
-                        *Tmp2++ = *Tmp;
+                        glyphCount = 2;
                     }
                 }
-                if (Tmp2 == buffer.get())
-                {
-                    LastChar = UNICODE_SPACE;
-                }
-                else
-                {
-#pragma prefast(suppress : 26001, "This is fine. Tmp2 has to have advanced or it would equal pBuffer.")
-                    LastChar = *(Tmp2 - 1);
-                }
 
-                if (LastChar == UNICODE_TAB)
+                for (;;)
                 {
-                    CursorPosition.x -= RetrieveNumberOfSpaces(sOriginalXPosition,
-                                                               pwchBufferBackupLimit,
-                                                               pwchBuffer - pwchBufferBackupLimit - 1);
-                    if (CursorPosition.x < 0)
+                    // We've already moved up if the cursor was in the first column so
+                    // we need to start off with overwriting the text with whitespace.
+                    // It wouldn't make sense to check the cursor position again already.
                     {
-                        CursorPosition.x = (coordScreenBufferSize.width - 1) / TAB_SIZE;
-                        CursorPosition.x *= TAB_SIZE;
-                        CursorPosition.x += 1;
-                        CursorPosition.y -= 1;
-
-                        // since you just backspaced yourself back up into the previous row, unset the wrap
-                        // flag on the prev row if it was set
-                        textBuffer.GetRowByOffset(CursorPosition.y).SetWrapForced(false);
+                        const auto previousColumn = pos.x;
+                        pos.x = textBuffer.GetRowByOffset(pos.y).NavigateToPrevious(previousColumn);
+
+                        RowWriteState state{
+                            .text = { &tabSpaces[0], 8 },
+                            .columnBegin = pos.x,
+                            .columnLimit = previousColumn,
+                        };
+                        textBuffer.Write(pos.y, textBuffer.GetCurrentAttributes(), state);

```

Comment:
This is interesting, but not necessarily _concerning_. We're writing the spaces with the _current_ attribute rather than the one that we're overstriking.

Reply:
Unfortunately, that's also the same as the current implementation. 🫤
But I think this is fine because the branch is pretty much exclusive for COOKED_READ and it only supports one attribute with which all text is written. (I think?)